src/libraries/System.Formats.Sse/src/System/Formats/Sse/SseEnumerable.cs

```diff
@@ -0,0 +1,546 @@
+                                _lineOffset += advance;
+                                _lineLength -= advance;
+                                continue;
+                            }
+                        }
+                        else
+                        {
+                            // Record the last position searched for a newline. The next time we search,
+                            // we'll search from here rather than from _lineOffset, in order to avoid searching
+                            // the same characters again.
+                            _lastSearchedForNewline = _lineOffset + _lineLength;
+                        }
+
+                        // We've processed everything in the buffer we currently can, so if we've already read EOF, we're done.
+                        if (_eof)
+                        {
+                            // Spec: "Once the end of the file is reached, any pending data must be discarded. (If the file ends in the middle of an
+                            // event, before the final empty line, the incomplete event is not dispatched.)"
+                            break;
+                        }
+
+                        // Read more data into the buffer.
+                        FillLineBuffer();
+                    }
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(_lineBuffer);
+                    if (_dataBuffer is not null)
+                    {
+                        ArrayPool<byte>.Shared.Return(_dataBuffer);
+                    }
+                }
+            }
+        }
+
+        /// <summary>Gets an enumerable of the server-sent events.</summary>
+        public IAsyncEnumerator<SseItem<T>> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            // Validate that the enumerable is only used for one enumeration.
+            ThrowIfNotFirstEnumeration();
+
+            return IterateEvents(cancellationToken);
+
+            async IAsyncEnumerator<SseItem<T>> IterateEvents(CancellationToken cancellationToken = default)
+            {
+                // Rent a line buffer. This will grow as needed. The line buffer is what's passed to the stream,
+                // so we want it to be large enough to reduce the number of reads we need to do when data is
+                // arriving quickly. (In debug, we use a smaller buffer to stress the growth and shifting logic.)
+                _lineBuffer = ArrayPool<byte>.Shared.Rent(DefaultArrayPoolRentSize);
+                try
+                {
+                    // Spec: "Event streams in this format must always be encoded as UTF-8".
+                    // Skip a UTF8 BOM if it exists at the beginning of the stream. (The BOM is defined as optional in the SSE grammar.)
+                    while (await FillLineBufferAsync(cancellationToken).ConfigureAwait(false) != 0 && _lineLength < Utf8Bom.Length) ;
```

Comment:
nit, but curious: there's a bunch of fairly detailed code shared across the sync/async versions of this method that seem to only differ by which variant of the FillLineBuffer method they call.  Wondering if there's a way to reuse these lines instead of copy?  Or would factoring to a shared method add perf overhead we seek to avoid?

Reply:
It's challenging to do well with the continues, breaks, and yield returns in there.